crates/cxx-qt-build/src/lib.rs

```diff
@@ -506,6 +507,26 @@ impl CxxQtBuilder {
             cc_builder_whole_archive.file(qtbuild.qrc(&qrc_file));
             cc_builder_whole_archive_files_added = true;
         }
+
+        // If we are using Qt 5 then write the std_types source
+        // This registers std numbers as a type for use in QML
+        //
+        // Note that we need this to be compiled into the whole_archive builder
+        // as they are stored in statics in the source.
+        if qtbuild.version().major == 5 {
+            let std_types_contents = include_str!("std_types_qt5.cpp");
+            let std_types_path = format!("{out_dir}/std_types_qt5.cpp");
+            let mut source =
+                File::create(&std_types_path).expect("Could not create std_types source");
+            write!(source, "{std_types_contents}").expect("Could not write std_types source");
+
+            cc_builder_whole_archive.file(&std_types_path);
+            // TODO: this line shouldn't be required,
```

Comment:
this shouldn't be required, needs investigating

Reply:
Actually in CMake builds double up whole_archive parts too.

Eg if you remove `qml/qml.qrc` from `CMakeLists.txt` and add it to the `build.rs` file, then it only works with cargo-only builds and not CMake builds.

So having `std_types.cpp` twice to get cargo-only and cmake-only builds both working is not that different and maybe OK for now.

Reply:
We maybe need to investigate separately why the `qt-static-initializers` is not included in the final binary with CMake. It kinda feels like either the other cc builder needs to link to it or CMake itself needs to link to it.

Reply:
Although looks like the `compile` method of `cc::Build` should be causing the `cargo:rustc-link-lib` to be set, so maybe something else is going wrong when CMake links this.

Reply:
Digging further into this it seems to be a general issue with main as stated before the same happens with qrc files.

Things which are built into `libqt-static-initializers.a` do not end up in the staticlib from cargo `libqml_minimal.a`, but things from `libcxx-qt-gen.a` do (even though the two libraries are built in very similar ways). Note that when cargo builds to an executable they do however remain.

It seems that cargo/rustc itself is doing something odd when compiling to a staticlib rather than an executable which means that the first library is lost / optimised out.

Note that swapping around the order of the `compile(...)` of the `cc::Build` causes the build to fail as it can't find things that are likely in the first library and missing.

So either we need to figure out how to cause cargo/rustc not to lose the non-last library or we need to move everything to being in one lib that is WHOLE_ARCHIVE

Reply:
So now this branch also fixes that qrc and QML modules wouldn't have worked with CMake builds if you put them only in the rust build script and not the CMakeLists.

The question now becomes is, should we just always build with whole_archive rather than having this split in the rust side ? As CMake builds link with whole_archive anyway?

Or can we improve the Rust side when building to a staticlib to not remove things?

Reply:
This doesn't work with cargo-only builds as you get duplicate implementations :see_no_evil: I'll try creating another branch which removes the double builder and tries that route...

Reply:
This has separate issues, i think this should be discussed in another issue / pull request as `main` suffers from this too.

Reply:
Split this into #597